### PR TITLE
remove hardcoded `device="cuda" ` to support more device

### PR DESCRIPTION
**Pull request by jikunshang**
Refer to #1948 , there are a lot of code use `cuda` as device, especially in tensor creation, which is not friendly to add other device support. This PR aims to refactor the code to leave some interface for better and easily add new device like `cpu` or `xpu`

### CODE REVIEW

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -446,51 +454,51 @@ def get_size_or_none(x: Optional[torch.Tensor]):
             py_data = receving_list[0]
             input_tokens = torch.empty(*py_data["input_tokens_size"],
                                        dtype=torch.long,
-                                       device="cuda")
+                                       device=self.device)
```
Comment (WoosukKwon) [inline vllm/worker/model_runner.py:457]: Can we use the `_set_default_torch_device` context manager here to not repeat `device=self.device`?

File: vllm/worker/model_runner.py [hunk 2]
```diff
@@ -147,18 +148,21 @@ def _prepare_prompt(
         input_tokens = _make_tensor_with_pad(input_tokens,
                                              max_prompt_len,
                                              pad=0,
-                                             dtype=torch.long)
+                                             dtype=torch.long,
+                                             device=self.device)
```
Comment (WoosukKwon) [inline vllm/worker/model_runner.py:152]: Can we use the `_set_default_torch_device` context manager here to not repeat `device=self.device`?

File: vllm/model_executor/model_loader.py [hunk 3]
```diff
@@ -58,11 +67,13 @@ def get_model(model_config: ModelConfig) -> nn.Module:
                 f"{supported_dtypes}")
         linear_method = quant_config.get_linear_method()
 
-    with _set_default_torch_dtype(model_config.dtype):
+    with _set_default_torch_dtype(
+            model_config.dtype), _set_default_torch_device(
+                model_config.device):
```
Comment (WoosukKwon) [inline vllm/model_executor/model_loader.py:72]: To my knowledge, python 3.8 does not support the nested with statement in a single line. Could you make this into two separate lines (and add a note so that people don't accidentally change it back to a single line)?

File: vllm/model_executor/model_loader.py [hunk 4]
```diff
@@ -21,6 +21,15 @@ def _set_default_torch_dtype(dtype: torch.dtype):
     torch.set_default_dtype(old_dtype)
 
 
+@contextlib.contextmanager
+def _set_default_torch_device(device: torch.device):
```
Comment (WoosukKwon) [inline vllm/model_executor/model_loader.py:25]: A dumb question: How is this different from `with torch.device()`?

File: tests/kernels/test_prefix_prefill.py [hunk 5]
```diff
@@ -39,20 +46,20 @@ def test_contexted_kv_attention(
                         num_heads,
                         head_size,
                         dtype=dtype,
-                        device='cuda')
+                        device=device)
```
Comment (WoosukKwon) [inline tests/kernels/test_prefix_prefill.py:49]: Can we just add `torch.set_default_device("cuda")` at the beginning of the function body? Same for other test functions.

File: vllm/worker/worker.py [hunk 6]
```diff
@@ -55,21 +55,24 @@ def __init__(
...
+
+            # This env var set by Ray causes exceptions with graph building.
+            os.environ.pop("NCCL_ASYNC_ERROR_HANDLING", None)
+            self.device = torch.device(f"cuda:{self.local_rank}")
+            torch.cuda.set_device(self.device)
+
+            _check_if_gpu_supports_dtype(self.model_config.dtype)
+        else:
+            raise RuntimeError("Not support device type: " +
+                               str(self.model_config.device))
```
Comment (WoosukKwon) [inline vllm/worker/worker.py:75]:
```suggestion
            raise RuntimeError("fNot support device type: {self.model_config.device}")
```

File: vllm/worker/worker.py [hunk 7]
```diff
@@ -55,21 +55,24 @@ def __init__(
...
-        os.environ["TORCH_NCCL_AVOID_RECORD_STREAMS"] = "1"
-
-        # This env var set by Ray causes exceptions with graph building.
-        os.environ.pop("NCCL_ASYNC_ERROR_HANDLING", None)
-        self.device = torch.device(f"cuda:{self.local_rank}")
-        torch.cuda.set_device(self.device)
-
-        _check_if_gpu_supports_dtype(self.model_config.dtype)
-
+        if self.model_config.device.type == "cuda":
```
Comment (WoosukKwon) [inline vllm/worker/worker.py:58]: How does it work with AMD GPUs?

File: vllm/worker/worker.py [hunk 8]
```diff
@@ -55,21 +55,24 @@ def __init__(
...
+            # torch.distributed.all_reduce does not free the input tensor   until
+            # the synchronization point. This causes the memory usage to grow
+            # as the number of all_reduce calls increases. This env var     disables
+            # this behavior.
+            # Related issue:
+            # https://discuss.pytorch.org/t/    cuda-allocation-lifetime-for-inputs-to-distributed-all-reduce/  191573
```
Comment (WoosukKwon) [inline vllm/worker/worker.py:64]:
```suggestion
            # https://discuss.pytorch.org/t/cuda-allocation-lifetime-for-inputs-to-distributed-all-reduce/191573
```

File: vllm/worker/worker.py [hunk 9]
```diff
@@ -55,21 +55,24 @@ def __init__(
...
-        self.device = torch.device(f"cuda:{self.local_rank}")
-        torch.cuda.set_device(self.device)
-
-        _check_if_gpu_supports_dtype(self.model_config.dtype)
-
+        if self.model_config.device.type == "cuda":
+            # torch.distributed.all_reduce does not free the input tensor   until
+            # the synchronization point. This causes the memory usage to grow
+            # as the number of all_reduce calls increases. This env var     disables
+            # this behavior.
```
Comment (WoosukKwon) [inline vllm/worker/worker.py:62]: Please remove the inserted spaces in the comment.

File: examples/offline_inference.py [hunk 10]
```diff
@@ -11,7 +11,7 @@
 sampling_params = SamplingParams(temperature=0.8, top_p=0.95)
 
 # Create an LLM.
-llm = LLM(model="facebook/opt-125m")
```
Comment (WoosukKwon) [inline examples/offline_inference.py:14]: Is this change necessary? Otherwise, can we revert it? This does not look nicer than the current one.

File: examples/offline_inference_with_prefix.py [hunk 11]
```diff
@@ -22,7 +22,7 @@
 sampling_params = SamplingParams(temperature=0.0)
 
 # Create an LLM.
-llm = LLM(model="facebook/opt-125m")
```
Comment (WoosukKwon) [inline examples/offline_inference_with_prefix.py:25]: ditto

File: vllm/config.py [hunk 12]
```diff
@@ -86,6 +87,7 @@ def __init__(
         self.quantization = quantization
         self.enforce_eager = enforce_eager
         self.max_context_len_to_capture = max_context_len_to_capture
+        self.device = torch.device(device)
```
Comment (WoosukKwon) [inline vllm/config.py:90]: I feel `device` is not a model's config but rather deserves its own config class like `DeviceConfig`. Could you make one?

File: vllm/engine/arg_utils.py [hunk 13]
```diff
@@ -202,6 +203,13 @@ def add_cli_args(
                             help='maximum context length covered by CUDA '
                             'graphs. When a sequence has context length '
                             'larger than this, we fall back to eager mode.')
+        parser.add_argument(
+            "--device",
+            type=str,
+            default="cuda",
+            choices=["cuda"],
+            help=
+            'device type for vLLM execution, supporting CUDA only currently.')
```
Comment (WoosukKwon) [inline vllm/engine/arg_utils.py:212]: I don't think we need this argument. I believe the device should be "detected" instead of being declared by the users.

**Review by WoosukKwon — CHANGES_REQUESTED** [review]
@jikunshang Thanks for submitting the PR! While the code looks good overall, I have two concerns:

1. I believe `device` should be automatically **detected** instead of being declared by users. This also aligns with our current design. While I don't know a good way to implement this is, I feel there should be a way to do it as long as the device is supported by PyTorch.
2. I believe `device` should not be an attribute of `ModelConfig`. Can we make a new config class like `DeviceConfig`?

File: vllm/model_executor/model_loader.py (hunk 3 above)
Comment (jikunshang) [reply vllm/model_executor/model_loader.py]: this change is unnecessary, revert back:)

File: vllm/model_executor/model_loader.py (hunk 4 above)
Comment (jikunshang) [reply vllm/model_executor/model_loader.py:25]: they are same, reverted

File: vllm/worker/worker.py (hunk 7 above)
Comment (jikunshang) [reply vllm/worker/worker.py]: amd GPU should have same behavior like cuda. https://pytorch.org/docs/stable/notes/hip.html

File: vllm/config.py (hunk 12 above)
Comment (jikunshang) [reply vllm/config.py]: Done.

File: vllm/engine/arg_utils.py (hunk 13 above)
Comment (jikunshang) [reply vllm/engine/arg_utils.py:212]: I think this may be necessary in the future. Image that if vllm have both cuda and cpu device supported, and user maybe want to run on cpu even if cuda path works on his environment.

**Comment by WoosukKwon** [comment]
Hi @jikunshang, could you resolve the merge conflicts? The PR looks good overall.

**Comment by jikunshang** [comment]
> Hi @jikunshang, could you resolve the merge conflicts? The PR looks good overall.

Sure. I have resolved conflicts. All tests should have been fixed.

**Review by liangfu — APPROVED** [review]

File: vllm/engine/arg_utils.py (hunk 13 above)
Comment (WoosukKwon) [reply vllm/engine/arg_utils.py:265]: Makes sense. Thanks for the explanation!

File: vllm/worker/model_runner.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/worker/model_runner.py:191]: After second thoughts, I found that explicitly specifying the devices would be better as we might mix CPU and accelerator in some cases.

File: vllm/worker/model_runner.py [hunk 14]
```diff
@@ -49,7 +50,10 @@ def __init__(
         # FIXME(woosuk): This is a hack to make the tests work. Refactor this.
         self.sliding_window = (model_config.get_sliding_window()
                                if model_config is not None else None)
-        self.device = torch.device(torch.cuda.current_device())
+        self.device_config = (device_config
+                              if device_config is not None else DeviceConfig())
+        self.device = self.device_config.device
```
Comment (WoosukKwon) [inline vllm/worker/model_runner.py:55]: BTW, I intentionally avoided using `torch.set_current_device`, since this can affect the user code when using `LLM` class.

**Review by WoosukKwon — APPROVED** [review]
@jikunshang LGTM! Thanks for submitting the PR and sorry for the delay in my second review.

While I think vLLM still has several `torch.cuda` calls, I believe this is a good first step towards supporting non-CUDA devices. Thanks for the great work!
